src/ds++/FMA.hh

```diff
@@ -26,41 +44,129 @@
+ *   sufficiently different.
+ * - Return the 'accurate' result.
+ */
+inline double fma_with_diagnostics(double const a, double const b,
+                                   double const c, std::string const &file,
+                                   uint32_t const line,
+                                   bool const abort_on_fail) {
+  double const accurate = fma(a, b, c);
+  double const fast = ((a) * (b) + c);
+  double const tol = 1.e-13;
+  std::ostringstream msg;
+  msg << "FMA accuracy error in " << file << " (" << line << ") "
+      << std::setprecision(std::numeric_limits<double>::max_digits10)
+      << "\n  accurate = " << accurate << "\n  fast = " << fast
+      << "\n  diff = " << fabs(accurate - fast)
+      << "\n  rdiff = " << 2.0 * fabs(accurate - fast) / fabs(accurate + fast)
+      << "\n  a = " << a << "\n  b = " << b << "\n  c = " << c << std::endl;
+  if (abort_on_fail) {
+    Insist(rtt_dsxx::soft_equiv(accurate, fast, tol), msg.str().c_str());
+  } else {
+    if (!rtt_dsxx::soft_equiv(accurate, fast, tol)) {
+      std::cout << msg.str();
+    }
+  }
 
+  return abort_on_fail ? fast : accurate;
+}
+} // namespace rtt_dsxx
+
+//----------------------------------------------------------------------------//
+/*!
+ * If \c FMA_FIND_DIFFS is defined, then hijack the whole FMA operation.
+ * Instead, perform the operation both ways and compare the result.  If the
+ * result is significantly different, throw an exception.  This is useful for
+ * finding operations that are sensitive to roundoff.
+ */
+#define FMA(a, b, c)                                                           \
+  rtt_dsxx::fma_with_diagnostics((a), (b), (c), __FILE__, __LINE__, true)
+#define FMA_ACCURATE(a, b, c)                                                  \
+  rtt_dsxx::fma_with_diagnostics((a), (b), (c), __FILE__, __LINE__, false)
+#else
+
+//----------------------------------------------------------------------------//
+//! HAVE_HARDWARE_FMA is set by config/platform_checks.cmake.
 #ifdef HAVE_HARDWARE_FMA
 
-/* This will be the default for new machines. Both versions should be equally
- * accurate, choose whatever is faster. FP_FAST_FMA is set by the compiler
+//----------------------------------------------------------------------------//
+/*!
+ * This will be the default for new machines. Both versions should be equally
+ * accurate, choose whatever is faster. \c FP_FAST_FMA is set by the compiler
  *
  * \ref http://en.cppreference.com/w/c/numeric/math/fma
+ *
+ * Working on LANL systems, it seems that \c FP_FAST_FMA is not set by the Intel
+ * compiler, so Intel will always pick the 2nd option.
+ *
+ * The intent of the \c FMA_ACCURATE version is to pick the more accurate
+ * multiply-plus-add operation. When hardware FMA is available, we just just the
+ * normal 'fast' version.
  */
 #ifdef FP_FAST_FMA
-#define FMA(a, b, c) fma(a, b, c)
+#define FMA(a, b, c) fma((a), (b), (c))
```

Comment:
It seems to me that this ifdef is switchted. Isn't ((a)*(b)+(c)) always the fastest?

Reply:
Also, why do we need FMA and FMA_ACCURATE, it looks like there definitions are always the same in these if blocks.

Reply:
* If `FP_FAST_FMA` is set, then the call to `fma(a,b,c)` is guaranteed to be the fastest computation (yes - faster than `a*b+c` and yes - it is confusing).  You can read the section at http://en.cppreference.com/w/c/numeric/math/fma for more details.
* I want the CPP macro `FMA_ACCURATE` to always be defined.  This allows us to always choose the more _accurate_ version, even when it is slower.  This is required in a few places to allow our tests to pass.  For machines that have hardware FMA, `FMA_ACCURATE == FMA`, but this isn't true for machines w/o hardware FMA.